src/allocator_global_pool.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+    Copyright (c) 2019-2020 Contributors as noted in the AUTHORS file
+
+    This file is part of libzmq, the ZeroMQ core engine in C++.
+
+    libzmq is free software; you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License (LGPL) as published
+    by the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    As a special exception, the Contributors give you permission to link
+    this library with independent modules to produce an executable,
+    regardless of the license terms of these independent modules, and to
+    copy and distribute the resulting executable under terms of your choice,
+    provided that you also meet, for each linked independent module, the
+    terms and conditions of the license of that module. An independent
+    module is a module which is not derived from or based on this library.
+    If you modify this library, you must extend this exception to your
+    version of the library.
+
+    libzmq is distributed in the hope that it will be useful, but WITHOUT
+    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+    FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+    License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "precompiled.hpp"
+#include "allocator_global_pool.hpp"
+
+#define ZMG_GLOBAL_POOL_INITIAL_BLOCK_SIZE (1024 * 256)
+
+zmq::allocator_global_pool_t::allocator_global_pool_t (
+  size_t initialMaximumBlockSize)
+{
+    _tag = 0xCAFEEBEC;
+    allocate_block (BytesToMsgBlock (initialMaximumBlockSize));
+}
+
+zmq::allocator_global_pool_t::~allocator_global_pool_t ()
+{
+    // deallocate all message classes
+    for (size_t i = 0U; i < _storage.size (); i++) {
+        for (size_t j = 0U; j < _storage[i].raw_data.size (); j++) {
+            operator delete (_storage[i].raw_data[j]);
+            _storage[i].raw_data[j] = NULL;
+        }
+    }
+    _tag = 0xdeadbeef;
+}
+
+bool zmq::allocator_global_pool_t::check_tag () const
+{
+    return _tag == 0xCAFEEBEC;
+}
+
+void zmq::allocator_global_pool_t::allocate_block (size_t bl)
+{
+    _storage_mutex.lock ();
+    size_t oldSize = _storage.size ();
+    if (oldSize <= bl) {
+        _storage.resize (bl + 1);
+        _free_list.resize (bl + 1);
```

Comment:
Is `allocate_block` supposed to be thread-safe? Because here _free_list is being mutated in a thread-unsafe manner.